### PR TITLE
Add support for manual fault tolerant

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for manual fault tolerant training.

There is currently 5 different modes defined as follows:

```py
class FaultTolerantTrainingModes(LightningEnum):

    DISABLED = "0"
    AUTOMATIC = "1"
    MANUAL = "2"
    AUTOMATIC_ELASTIC = "3"
    MANUAL_ELASTIC = "4"
```

In order to prevent too many patching, this PR introduces a `_SingleProcessDataLoaderIterStateful` and  `_MultiProcessingDataLoaderIterStateful`.

In further PR, the code between automatic and manual would be further merged.

Parts of #10431

TODO:
- [x] Validate it works with 1 worker
- [x] Validate it works with 2 workers

To be merged after
* #10674

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @justusschock @awaelchli

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -22,12 +22,29 @@
...
+)
 
 import pytorch_lightning as pl
 from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
-from pytorch_lightning.utilities.imports import _fault_tolerant_training
+from pytorch_lightning.utilities.imports import _fault_tolerant_training, _fault_tolerant_training_mode
+
+
+class _FaulTolerantAbstract(abc.ABC):
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/auto_restart.py:39]: Unsure what @awaelchli or @carmocca think but my opinion is that this is confusing. We have an API being enforced internally that the user has to be aware of, but there is no way to know this other than documentation.

I think it's better we take the OO approach by enforcing the user to inherit from a particular class to make the sampler/dataloader stateful (either via inheritance or multiple inheritance)

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: Changed to functional API.

File: pytorch_lightning/utilities/enums.py [hunk 2]
```diff
@@ -256,3 +256,28 @@ def interactive_compatible_types() -> List["_StrategyType"]:
...
         return self in _StrategyType.interactive_compatible_types()
+
+
+class FaultTolerantTrainingModes(LightningEnum):
+
+    DISABLED = "0"
+    AUTOMATIC = "1"
+    MANUAL = "2"
+    AUTOMATIC_ELASTIC = "3"
+    MANUAL_ELASTIC = "4"
```
Comment (awaelchli) [inline pytorch_lightning/utilities/enums.py:267]: these are not supported yet, right? They should not be included

File: pytorch_lightning/trainer/data_loading.py [hunk 3]
```diff
@@ -293,7 +293,13 @@ def _get_dataloader_init_kwargs(
...
             dl_kwargs["sampler"] = None
 
-        if _fault_tolerant_training():
+        dl_kwargs = TrainerDataLoadingMixin._prepare_fault_tolerance(dl_kwargs)
+
+        return dl_kwargs
+
+    @staticmethod
+    def _prepare_fault_tolerance(dl_kwargs: Dict) -> Dict:
+        if _fault_tolerant_training_mode().is_automatic:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:302]: The name is misleading if the body only gets executed for some modes. I propose:

```py
if _fault_tolerant_training_mode().is_automatic:
    dl_kwargs["dataset"] = _capture_dataset(dl_kwargs["dataset"])


def _capture_dataset(...):
    ...
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -351,6 +356,11 @@ def __next__(self) -> Dict[str, Any]:
         return next(self.iter_data)
 
 
+def is_obj_stateful(obj: Any) -> bool:
+    """In order to be stateful, an object should implement a ``state_dict`` and ``load_state_dict`` method."""
+    return getattr(obj, "state_dict", None) is not None and getattr(obj, "load_state_dict", None) is not None
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:361]: use hasattr? 
it may need a callable check?

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -455,12 +465,68 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
...
+    if mode.is_automatic:
+        metadata = dataset.state_dict()
+    else:
+        state_dict_fn = getattr(dataset, "state_dict", None)
+        info = get_worker_info()
+        worker_id = info.id if info else 0
+        if state_dict_fn:
+            metadata = state_dict_fn()
+            if worker_id not in metadata:
+                raise MisconfigurationException("The state_dict needs to be indexed by `worker_id`")
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:481]:
```suggestion
                raise MisconfigurationException("The state_dict returned by {the name/class of the object} needs to be indexed by `worker_id` integer keys.")
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 6]
```diff
@@ -570,4 +598,146 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
...
+            for worker_id in state_dict["state"].keys()
+        }
+
+        if num_workers > 0:
+            # remap states to worker ids starting at 0
+            next_worker_id = latest_worker_id + 1
+            old_to_new_worker_id_map = [((next_worker_id + i) % num_workers, i) for i in range(num_workers)]
+            dataset_state = {
+                new_id: dataset_state[old_id] for old_id, new_id in old_to_new_worker_id_map if old_id in dataset_state
+            }
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:631]: this code is duplicated from CaptureMapDataset. It already gets done in load_state_dict() of CaptureMapDataset which is what gets called in the line below. 
So there is probably a bug here because it will rotate the states twice.

File: pytorch_lightning/utilities/enums.py [hunk 7]
```diff
@@ -256,3 +256,28 @@ def interactive_compatible_types() -> List["_StrategyType"]:
     def is_interactive_compatible(self) -> bool:
         """Returns whether self is interactive compatible."""
         return self in _StrategyType.interactive_compatible_types()
+
+
+class FaultTolerantTrainingModes(LightningEnum):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/enums.py:261]: Done !

File: pytorch_lightning/utilities/enums.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/enums.py:267]: No, done !

File: pytorch_lightning/trainer/data_loading.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:302]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:361]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:481]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:631]: Fault manual, there is no CaptureDataset been used, meaning we need to rotate the indices. I extracted this logic into its own function.

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: Done !

File: pytorch_lightning/utilities/enums.py [hunk 8]
```diff
@@ -256,3 +256,22 @@ def interactive_compatible_types() -> List["_StrategyType"]:
     def is_interactive_compatible(self) -> bool:
         """Returns whether self is interactive compatible."""
         return self in _StrategyType.interactive_compatible_types()
+
+
+class FaultTolerantTrainingMode(LightningEnum):
+
+    DISABLED = "0"
+    AUTOMATIC = "1"
+    MANUAL = "2"
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:265]: Either these should have actual strings or you should not subclass `LightningEnum`

```suggestion
    DISABLED = "disabled"
    AUTOMATIC = "automatic"
    MANUAL = "manual"
```

File: pytorch_lightning/utilities/enums.py [hunk 9]
```diff
@@ -256,3 +256,22 @@ def interactive_compatible_types() -> List["_StrategyType"]:
...
+
+    @property
+    def is_enabled(self) -> bool:
+        return self != FaultTolerantTrainingMode.DISABLED
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:269]:
```suggestion
        return self is not FaultTolerantTrainingMode.DISABLED
```

**Review by carmocca — COMMENTED** [review]
Any chance this PR can be split in smaller pieces?
